app/assets/javascripts/admin/addon/services/admin-search-data-source.js

```diff
@@ -271,22 +277,42 @@ export default class AdminSearchDataSource extends Service {
         ? MAX_TYPE_RESULT_COUNT_LOW
         : MAX_TYPE_RESULT_COUNT_HIGH;
 
+    const labelStartRegex = new RegExp(`^${escapedFilterRegExp}`, "i");
+    const exactKeywordRegexes = escapedFilterRegExp
+      .split(" ")
+      .map((keyword) => new RegExp(`(${keyword})\\b`, "i"));
+    const fallbackRegex = new RegExp(`${escapedFilterRegExp}`, "i");
+
     opts.types.forEach((type) => {
-      let typeItemCount = 0;
+      const typeResults = [];
       this[`${type}DataSourceItems`].forEach((dataSourceItem) => {
-        // TODO (martin) There is likely a much better way of doing this matching
-        // that will support fuzzy searches, for now let's go with the most basic thing.
+        dataSourceItem.score = 0;
+
+        if (dataSourceItem.label.match(labelStartRegex)) {
+          dataSourceItem.score =
+            dataSourceItem.score + SEARCH_SCORES.labelStart;
+        }
         if (
-          dataSourceItem.keywords.match(escapedFilterRegExp) &&
-          typeItemCount <= perTypeLimit
+          exactKeywordRegexes.every((regex) => {
+            return dataSourceItem.label.match(regex);
+          })
         ) {
-          filteredResults.push(dataSourceItem);
-          typeItemCount++;
+          dataSourceItem.score =
+            dataSourceItem.score + SEARCH_SCORES.exactKeyword;
+        }
+        if (dataSourceItem.keywords.match(fallbackRegex)) {
+          dataSourceItem.score = dataSourceItem.score + SEARCH_SCORES.fallback;
```

Comment:
(ultra-nit) can be shortened a bit:

```suggestion
          dataSourceItem.score += SEARCH_SCORES.fallback;
```